text/0000-feature-metadata-doc-cfg.md

```diff
@@ -0,0 +1,334 @@
+   `--config-toml`; no parsing of these files is necessary.
+
+Cargo should use the following precedence (later items take priority over
+earlier items):
+
+1. Workspace `Cargo.toml`
+2. Workspace root `rustdoc.toml`
+3. Crate `Cargo.toml`
+4. Crate root `rustdoc.toml`
+
+`rustdoc` will be in charge of handling configuration merging. This should
+create an intuitive layering of global options and overrides while keeping
+`rustdoc` and `Cargo` reasonably separate.
+
+## Changes to `rustdoc`
+
+`rustdoc` must be aware of two new arguments: `--config-json` and
+`--config-file`. `--config-json` accepts either a JSON file path or a JSON
+string, `--config-file` accepts a path to a TOML file. The JSON and TOML
+share an identical schema to what is shown above:
+
+```json5
+{
+    "html-logo-url": "https://example.com/logo.jpg",
+    "issue-tracker-base-url": "https://github.com/rust-lang/rust/issues/",
+    "features": {
+        "foo": [],
+        "bar": { "doc": "simple docstring here", "requires": ["foo"] },
+        "baz": { "public": false, "requires": ["bar"] },
+        "qux": {
+            "doc": "# corge\n\nThis could be a longer description of this feature\n",
+            "requires": ["bar", "baz"]
+        }
+    }
+}
+```
+
+Spans can also be specified for JSON files for better diagnostics. It is
+expected that Cargo could maintain span information for data extracted from
+`Cargo.toml`, but it is not required that other build systems or handwritten
+configuration provide this information. This is also not required for a minimum
+viable product.
+
+```json5
+{
+    // Indicate the source file that spans are based on
+    "_root-span-path": "/path/to/Cargo.TOML",
+    "html-logo-url": {
+        "data": "https://example.com/logo.jpg",
+        "start": 100,
+        "end": 123
+    },
+    "features": {
+        "data": {
+            "foo": { "data": [], "start": 10, "end": 15 },
+            "bar": {
+                "data": {
+                    "doc": { "data": "simple docstring here", "start": 15, "end": 20 },
+                    "requires": { "data": ["foo"], "start": 20, "end": 25 }
+                },
+                "start": 15,
+                "end": 30
+            }
+        },
+        "start": 10,
+        "end": 100
+    }
+}
+```
+
+`rustdoc` should start with a default configuration and update/overwrite it with
+each `--config-file` or `--config-json` argument. Configuration specified in
+rust source files (e.g. `#![doc(html_favicon_url ="foo.com/favicon")]`) take the
+highest priority.
+
+# Drawbacks
+
+[drawbacks]: #drawbacks
+
+- This adds complexity to `rustdoc`. In this case, it does seem like it is
+  justified.
+- This adds noise to the Cargo manifest that is not relevant to Cargo itself.
+  This RFC seeks to provide a good middle ground: lengthy or complex
+  configuration or feature descriptions can exist in a separate `rustdoc.toml`,
+  but a separate file also isn't required for simple configuration.
```

Comment:
While #3389 doesn't have tool config yet, it does overlap in discussion around tool config on how to handle tools getting `Cargo.toml` configuration

Reply:
I listed that as a related issue in https://github.com/rust-lang/rfcs/pull/3421